StreamChatClient_v3/Controllers/ChannelController.swift

```diff
@@ -5,78 +5,300 @@
-        self.channelId = channelId
-        self.worker = worker
-        self.viewContext = viewContext
+    private(set) lazy var messagesObserver: ListDatabaseObserver<MessageModel<ExtraData>, MessageDTO> = {
+        let observer = ListDatabaseObserver(context: self.client.databaseContainer.viewContext,
+                                            fetchRequest: MessageDTO.messagesFetchRequest(for: self.channelQuery.cid),
+                                            itemCreator: MessageModel<ExtraData>.init)
+        observer.onChange = { [unowned self] changes in
+            self.delegateCallback {
+                $0?.channelController(self, didUpdateMessages: changes)
+            }
+        }
+        
+        return observer
+    }()
+    
+    private let environment: Environment
+    
+    /// Creates a new `ChannelController`
+    /// - Parameters:
+    ///   - channelQuery: channel query for observing changes
+    ///   - client: The `Client` this controller belongs to.
+    ///   - environment: Envrionment for this controller.
+    init(channelQuery: ChannelQuery<ExtraData>, client: Client<ExtraData>, environment: Environment = .init()) {
+        self.channelQuery = channelQuery
+        self.client = client
+        self.environment = environment
     }
     
-    public func startUpdating() {
-        try! fetchResultsController.performFetch()
+    /// Starts updating the results.
+    ///
+    /// 1. **Synchronously** loads the data for the referenced objects from the local cache. These data are immediately available in
+    /// the `channel` property of the controller once this method returns. Any further changes to the data are communicated
+    /// using `delegate`.
+    ///
+    /// 2. It also **asynchronously** fetches the latest version of the data from the servers. Once the remote fetch is completed,
+    /// the completion block is called. If the updated data differ from the locally cached ones, the controller uses the `delegate`
+    /// methods to inform about the changes.
+    ///
+    /// - Parameter completion: Called when the controller has finished fetching remote data. If the data fetching fails, the `error`
+    /// variable contains more details about the problem.
+    public func startUpdating(_ completion: ((_ error: Error?) -> Void)? = nil) {
+        do {
+            try channelObserver.startObserving()
+            try messagesObserver.startObserving()
+        } catch {
+            log.error("Failed to perform fetch request with error: \(error). This is an internal error.")
+            callback { completion?(ClientError.FetchFailed()) }
+            return
+        }
         
-        guard let channel = fetchResultsController.fetchedObjects?.first
-            .map(ChannelModel<ExtraData>.create(fromDTO:)) else { return }
-        delegate?.channelController(self, didUpdateChannel: channel)
+        state = .active
```

Comment:
Interesting choice... why do we need the state, exactly? Just to check if this function has been called?

Reply:
Yes. I also have this in #356. I think it will be useful in the future, too, for example for performance optimizations - we don't need certain things if the observer is not active. If we see it's not needed, we can delete later.